Add Wiki section to the documentation (#84)
davenverse/shellfish: docs/wiki/file_reading_writing.md

# Reading, writing, and appending 

> This library comes with three different functions for reading and writing operations, `read`, `write` and `append`, each with four different variants: Standalone, `Bytes`, `Lines` and `As`, with these variants you will be able to work with the file and/or its contents as a UTF-8 string or with a custom `java.nio.charset.Charset`, as bytes, line by line, and with a custom codec respectively. 

Reviewer: ```suggestion
Shellfish comes with three different functions for reading and writing operations, `read`, `write` and `append`, each with four different variants: Standalone, `Bytes`, `Lines` and `As`, with these variants you will be able to work with the file and/or its contents as a UTF-8 string or with a custom `java.nio.charset.Charset`, as bytes, line by line, and with a custom codec respectively. 
```

```scala mdoc:invisible
// This section adds every import to the code snippets

import cats.effect.IO
import cats.syntax.all.*

import fs2.Stream
import fs2.io.file.{Path, Files}

import io.chrisdavenport.shellfish
import shellfish.syntax.path.*
import shellfish.FilesOs

val path = Path("testdata/dummy.something")
```

## Reading

One of the most common operations in scripting is reading a file, that is why you can read a file in different ways: 

### `read`

This function loads the whole file as a string in memory using UTF-8 encoding.

@:select(api-style)

@:choice(syntax)

```scala mdoc:compile-only
import cats.syntax.all.*

import shellfish.syntax.path.*

val path = Path("link/to/your/file.txt")

path.read >>= IO.println
```

@:choice(static)

```scala scala mdoc:compile-only
import cats.syntax.all.* 

import shellfish.FilesOs

val path = Path("link/to/your/file.txt")

FilesOs.read(path) >>= IO.println
```

@:choice(fs2)

```scala mdoc:compile-only
import fs2.io.file.Files

val path = Path("link/to/your/file.txt")

Files[IO].readUtf8(path).evalMap(IO.println).compile.drain
```

@:@

> If UTF-8 is not your favorite flavour, you can also use a custom `java.nio.charset.Charset` to decode your file:

Reviewer: ```suggestion
If you need to handle non UTF-8 files, you can also use a custom `java.nio.charset.Charset`:
```

@:select(api-style)

@:choice(syntax)

```scala 3 mdoc:compile-only
import java.nio.charset.StandardCharsets

path.read(StandardCharsets.UTF_16)
```

@:choice(static)

```scala 3 mdoc:compile-only
import java.nio.charset.StandardCharsets

FilesOs.read(path, StandardCharsets.UTF_16)
```

@:choice(fs2)

```scala mdoc:compile-only
import fs2.text.decodeWithCharset
import java.nio.charset.StandardCharsets

Files[IO].readAll(path)
  .through(decodeWithCharset(StandardCharsets.UTF_16))
  .compile
  .string
```

@:@

### `readBytes`

> Reads the file as a `ByteVector`, useful when working with binary data:

Reviewer: Maybe add the link to scodec's documentation for `ByteVector`?

@:select(api-style)

@:choice(syntax)

```scala mdoc:compile-only

> path.readBytes.map(_.rotateLeft(2))

Reviewer: Add a one line comment explaining what `rotateLeft` does

```

@:choice(static)

```scala mdoc:compile-only
FilesOs.readBytes(path).map(_.rotateLeft(2))
```

@:choice(fs2)

```scala mdoc:compile-only
import scodec.bits.ByteVector

Files[IO].readAll(path).compile.to(ByteVector).map(_.rotateLeft(2))
```

@:@

### `readLines`

> Similar to `read` as it reads the file as a UTF-8 string, but stores each line of the file on a `List`:

Reviewer: ```suggestion
Similar to `read` as it reads the file as a UTF-8 string, but stores each line of the file in a `List`:
```

@:select(api-style)

@:choice(syntax)

```scala mdoc:compile-only
for
  lines <- path.readLines

>   _     <- IO(lines.foreach(println))

Reviewer: !!!
```suggestion
  _     <- lines.traverse_(IO.println)
```

yield ()
```

@:choice(static)

```scala mdoc:compile-only
for
  lines <- FilesOs.readLines(path)
  _     <- IO(lines.foreach(println))
yield ()
```

@:choice(fs2)

```scala mdoc:compile-only
Files[IO].readUtf8Lines(path).evalMap(IO.println).compile.drain
```

@:@

### `readAs`

> This method reads the contents of the file given a `Codec[A]` in scope, useful when you want to convert a file into a custom type `A`:

Reviewer: ```suggestion
This method reads the contents of the file given a `Codec[A]`. Useful when you want to convert a binary file into a custom type `A`:
```

@:select(api-style)

@:choice(syntax)

```scala mdoc:compile-only
import scodec.Codec

case class Coordinates(x: Double, y: Double) derives Codec

path.readAs[Coordinates].map(coord => coord.x + coord.y)
```

@:choice(static)

```scala mdoc:compile-only
import scodec.Codec

case class Coordinates(x: Double, y: Double) derives Codec

FilesOs.readAs[Coordinates](path).map(coord => coord.x + coord.y)
```

@:choice(fs2)

```scala mdoc:compile-only
import scodec.Codec
import fs2.interop.scodec.StreamDecoder

case class Coordinates(x: Double, y: Double) derives Codec

Files[IO].readAll(path)
  .through(StreamDecoder.many(summon[Codec[Coordinates]]).toPipeByte)
  .map(coord => coord.x + coord.y)
  .compile
  .drain
```

@:@

## Writing

> You can also overwrite the contents of a file using the `write` method and any of its variants. Just to note, if the file does not exist, the file will be created automatically:

Reviewer 2: ```suggestion
You can also overwrite the contents of a file using the `write` method and any of its variants. If the the file does not exist, it will be created automatically:
```

```

@:choice(static)

```scala mdoc:compile-only
val path = Path("path/to/write.txt")

val poem = 
  """Music, thou queen of heaven, care-charming spell,
    |That strik'st a stillness into hell;
    |Thou that tam'st tigers, and fierce storms, that rise,
    |With thy soul-melting lullabies;
    |Fall down, down, down, from those thy chiming spheres
    |To charm our souls, as thou enchant'st our ears.
    |
    |— Robert Herrick""".stripMargin

FilesOs.write(path, poem)
```

@:choice(fs2)

```scala mdoc:compile-only
val path = Path("path/to/write.txt")

val poem = 
  """Music, thou queen of heaven, care-charming spell,
    |That strik'st a stillness into hell;
    |Thou that tam'st tigers, and fierce storms, that rise,
    |With thy soul-melting lullabies;
    |Fall down, down, down, from those thy chiming spheres
    |To charm our souls, as thou enchant'st our ears.
    |
    |— Robert Herrick""".stripMargin

Stream.emit(poem)
  .through(Files[IO].writeUtf8(path))
  .compile
  .drain
```

@:@

The default charset can also be changed for encoding when writing strings:

@:select(api-style)

@:choice(syntax)

```scala 3 mdoc:compile-only
import java.nio.charset.StandardCharsets

> val message = "Imagine writing Java 😭"

Reviewer: This is offensive, change it to another string

path.write(message, StandardCharsets.US_ASCII)
```

@:choice(static)

```scala 3 mdoc:compile-only
import java.nio.charset.StandardCharsets

val message = "Imagine writing Java 😭"

FilesOs.write(path, message, StandardCharsets.US_ASCII)
```

@:choice(fs2)

```scala mdoc:compile-only
import fs2.text.encode
import java.nio.charset.StandardCharsets

val message = "Imagine writing Java 😭"

Stream.emit(message)
  .through(encode(StandardCharsets.US_ASCII))
  .through(Files[IO].writeAll(path))
  .compile
  .drain
```

@:@

### `writeBytes`

With this method you can write bytes directly to a binary file. The contents of the file require to be in form of a `scodec.bits.ByteVector`:

@:select(api-style)

@:choice(syntax)

```scala mdoc:compile-only
import scodec.bits.*

> val niceBytes = hex"BadBabe"

Reviewer: Use `deadbeef` or similar, non `badbabe`
